Fix the ordering of variable initialisation in goto-harness

xbauch opened the pull request:
When using the `memory-snapshot` harness.

See https://github.com/diffblue/cbmc/issues/4978 for the problem.

Previously we were using the `refers-to` relation (`a` refers to `b` if the value `a` should be initialised with contains -- as a subexpression -- the symbol `b`) as if it was a linear ordering between the snapshot symbols. That is, we called `std::stable_sort` with this relation which is incorrect.

Now we select one linearisation of the partial order (given by `refers-to`) to be the initilisation ordering. Note that we ignore cycles, or more precisely in the presence of a cycle we start from the first visited symbol. This is correct, because for

```
struct A { A*next; };
A a1;
A a2;
a1.next = &a2;
a2.next = &a1;
```
the snapshot is

```
a1 = {.next=&a2}
a2 = {.next=&a1}
```

And the order is irrelevant. There is also a test-case which was failing before.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: e034622be65d7b4a06a7f058bc846af1992abac9).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/126261769

NlightNFotis commented on src/goto-harness/memory_snapshot_harness_generator.cpp:
```diff
@@ -207,16 +207,16 @@ size_t memory_snapshot_harness_generatort::pointer_depth(const typet &t) const
 }
 
-bool memory_snapshot_harness_generatort::refers_to(
+void memory_snapshot_harness_generatort::collect_references(
```
While I understand the need for the change for this, I feel slightly uneasy about the change to the return value (which provides some sort of feedback) to a void return value.

hannes-steffenhagen-diffblue commented:
I think the first commit here should probably just have the entire description of this PR in it

NlightNFotis reviewed (approved):
LGTM

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -275,11 +275,68 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
+
+    void sort(
+      const std::vector<std::pair<Key, Value>> &input,
+      std::vector<std::pair<Key, Value>> &output)
```
Is there a reason this can't do an in-place sort?

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -275,11 +275,68 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
+      const Key &key,
+      const std::map<Key, Value> &input,
+      std::vector<std::pair<Key, Value>> &output)
+    {
```
This seems to be basically doing a DFS. Might be worth just renaming the function to reflect that.

It could just have an API like

```
template<typename Value, typename GetNodeFromValue, typename GetNeighbors, typename OnFirstVisit>
void depth_first_search(Value &&entry_node, GetKeyFromValue &&get_key_from_value, GetNeighbors &&getNeighbors, OnFirstVisit &&onFirstVisit);
```

and then you could use it above like
```
auto get_key_from_value = [](const std::pair<Key, Value>& kv_pair) { return kv_pair.first; };
auto get_neighbors = [&](const Key& key) { return por.at(key); }
auto on_first_visit = [&](const Value& value) { output.push_back(value); }
for(const auto &item : input) {
  dfs(item, get_key_from_value, get_neighbors, on_first_visit);
}
```

(Note that I've changed `Value` to what was previously `std::pair<Key,Value>`, and instead there's a separate function for getting the key from the value).

:exclamation: This is just a suggestion, I'm not saying you have to do it exactly this way obviously, but in any case I think it's worth emphasising that "recurse then insert" is basically doing DFS in the code, either by renaming the function or mentioning it in a comment

xbauch replied:
I went half-way. Given that the ordering is on `keys` only, the structure is only templated by `Key` and the `sort`, `dfs` functions and templated by some `T`.

hannes-steffenhagen-diffblue commented on src/goto-harness/memory_snapshot_harness_generator.h:
```diff
@@ -275,11 +275,68 @@ class memory_snapshot_harness_generatort : public goto_harness_generatort
+    const std::map<Key, std::set<Key>> &por;
+
+    std::set<Key> inserted;
+    std::set<Key> seen;
```
Note that these make `sort` a destructive method, i.e. an instance becomes unusable after `sort` is first called. Might be better to just pass these on as parameters, or loopify the recursion and just keep these as local variables

xbauch replied:
I wrapped the `dfs` in a `cleaner` function.

hannes-steffenhagen-diffblue reviewed (approved):
Looks mostly good to me, but I think the partial order sort class could use some cleanup.

petr-bauch committed:
Fix the ordering of variable initialisation

by linearising the preorder induced by the symbol values referring to other
symbols.

Previously we were using the `refers-to` relation (`a` refers to `b` if the
value `a` should be initialised with contains -- as a subexpression -- the
symbol `b`) as if it was a linear ordering between the snapshot symbols. That
is, we called std::stable_sort with this relation which is incorrect.

Now we select one linearisation of the preorder (given by `refers-to`) to be the
initilisation ordering. Note that we ignore cycles, or more precisely in the
presence of a cycle we start from the first visited symbol.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: 373947eae08aaad5bb1da3fa81af594f85ed0fbb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/129287253
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: a4df02c57215a52841fad69cabfd1b4711495695).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/129292085
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

petr-bauch committed:
Add a test for the new preorder linearisation

when setting the order in which to initialise the variables from a snapshot.
This test is designed to be failing slightly more often than the
`dynamic-array-int` test by inserting extra variables in between the
preordered ones. They cannot be named `array` because that irep-name already
exists.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: fbb40a044ee56870d6f4d8007a0ec62d173f9e2c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/134848870
